File: truss-compiler/src/test/java/truss/compiler/parser/ParsingFixture.java
package truss.compiler.parser;

import truss.compiler.Closeable;
import truss.compiler.FixtureBase;
import truss.compiler.MessageCollection;
import truss.compiler.MessageCollectionScope;
import truss.compiler.printing.SyntaxPrinter;
import truss.compiler.printing.TextPrinter;
import truss.compiler.syntax.CompilationUnitSyntax;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.io.StringWriter;
import java.io.Writer;

import static org.junit.Assert.assertEquals;

@RunWith(JUnit4.class)
public class ParsingFixture extends FixtureBase {
    @Test
    public void basicSyntax() throws Exception {
        test("BasicSyntax.t3s");
    }

    @Test
    public void namespace() throws Exception {
        test("Namespace.t3s");
    }

    @Test
    public void attributes() throws Exception {
        test("Attributes.t3s");
    }

    @Test
    public void types() throws Exception {
        test("Types.t3s");
    }

    @Test
    public void typeMembers() throws Exception {
        test("TypeMembers.t3s");
    }

    @Test
    public void statements() throws Exception {
        test("Statements.t3s");
    }

    @Test
    public void names() throws Exception {
        test("Names.t3s");
    }

    @Test
    public void expressions() throws Exception {
        test("Expressions.t3s");
    }

    @Override
    public void test(String name, String fileName, String code, String expected) throws Exception {
        String actual;

        try (Writer writer = new StringWriter()) {
            MessageCollection messages = new MessageCollection();
            CompilationUnitSyntax compilationUnit;

            try (Closeable ignored = MessageCollectionScope.create(messages)) {
                compilationUnit = parse(name, code);
            }

            if (messages.hasMessages()) {
                actual = messages.toString().trim();
            } else {
                compilationUnit.accept(new SyntaxPrinter(new TextPrinter(writer)));
                actual = writer.toString().trim();
            }
        }

        assertEquals(String.format("Parse of '%s' failed", name), expected, actual);
    }
}

File: Truss.Compiler.Test/Parser/ParsingFixture.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using NUnit.Framework;
using Truss.Compiler.Printing;
using Truss.Compiler.Syntax;

namespace Truss.Compiler.Test.Parser {
    [TestFixture]
    public class ParsingFixture : FixtureBase {
        [Test]
        public void BasicSyntax() {
            Test("BasicSyntax.trs");
        }

        [Test]
        public void Namespace() {
            Test("Namespace.trs");
        }

        [Test]
        public void Attributes() {
            Test("Attributes.trs");
        }

        [Test]
        public void Types() {
            Test("Types.trs");
        }

        [Test]
        public void TypeMembers() {
            Test("TypeMembers.trs");
        }

        [Test]
        public void Statements() {
            Test("Statements.trs");
        }

        [Test]
        public void Names() {
            Test("Names.trs");
        }

        [Test]
        public void Expressions() {
            Test("Expressions.trs");
        }

        public override void Test(String name, String fileName, String code, String expected) {
            String actual;

            using (var writer = new StringWriter()) {
                var messages = new MessageCollection();
                CompilationUnitSyntax compilationUnit;

                using (MessageCollectionScope.Create(messages)) {
                    compilationUnit = Parse(name, code);
                }

                if (messages.HasMessages) {
                    actual = messages.ToString().Trim();
                } else {
                    compilationUnit.Accept(new SyntaxPrinter(new TextPrinter(writer)));
                    actual = writer.ToString().Trim();
                }
            }

            Assert.AreEqual(expected, actual, String.Format("Parse of '{0}' failed", name));
        }
    }
}
